File: VB/PersonalOrganizer/Common/DataModel/DesignTimeRepository.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports DevExpress.Mvvm
Imports Common.Utils
Namespace Common.DataModel
    Public MustInherit Class DesignTimeRepository(Of TEntity As Class, TPrimaryKey)
        Inherits DesignTimeReadOnlyRepository(Of TEntity)
        Implements IRepository(Of TEntity, TPrimaryKey)
        Private _getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey))
        Private _getPrimaryKeyFunction As Func(Of TEntity, TPrimaryKey)
        Private _setPrimaryKeyAction As Action(Of TEntity, TPrimaryKey)
        Public Sub New(ByVal getPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)), Optional ByVal setPrimaryKeyAction As Action(Of TEntity, TPrimaryKey) = Nothing)
            Me._getPrimaryKeyExpression = getPrimaryKeyExpression
            Me._setPrimaryKeyAction = setPrimaryKeyAction
            If Me._setPrimaryKeyAction Is Nothing Then
                Me._setPrimaryKeyAction = getPrimaryKeyExpression.SetValueAction().Compile()
            End If
        End Sub
        Protected Overridable Function CreateCore() As TEntity
            Return DesignTimeHelper.CreateDesignTimeObject(Of TEntity)()
        End Function
        Protected Overridable Function FindCore(ByVal key As TPrimaryKey) As TEntity
            Throw New InvalidOperationException()
        End Function
        Protected Overridable Sub RemoveCore(ByVal entity As TEntity)
            Throw New InvalidOperationException()
        End Sub
        Protected Overridable Function ReloadCore(ByVal entity As TEntity) As TEntity
            Throw New InvalidOperationException()
        End Function
        Protected Overridable Function GetPrimaryKeyCore(ByVal entity As TEntity) As TPrimaryKey
            If _getPrimaryKeyFunction Is Nothing Then
                _getPrimaryKeyFunction = _getPrimaryKeyExpression.Compile()
            End If
            Return _getPrimaryKeyFunction(entity)
        End Function
        Protected Overridable Sub SetPrimaryKeyCore(ByVal entity As TEntity, ByVal key As TPrimaryKey)
            _setPrimaryKeyAction(entity, key)
        End Sub
        Private Function Find(ByVal key As TPrimaryKey) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Find
            Return FindCore(key)
        End Function
        Private Sub Remove(ByVal entity As TEntity) Implements IRepository(Of TEntity, TPrimaryKey).Remove
            RemoveCore(entity)
        End Sub
        Private Function Create() As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Create
            Return CreateCore()
        End Function
        Private Function Reload(ByVal entity As TEntity) As TEntity Implements IRepository(Of TEntity, TPrimaryKey).Reload
            Return ReloadCore(entity)
        End Function
        Private ReadOnly Property GetPrimaryKeyExpression As Expression(Of Func(Of TEntity, TPrimaryKey)) Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKeyExpression
            Get
                Return _getPrimaryKeyExpression
            End Get
        End Property
        Private Function GetPrimaryKey(ByVal entity As TEntity) As TPrimaryKey Implements IRepository(Of TEntity, TPrimaryKey).GetPrimaryKey
            Return GetPrimaryKeyCore(entity)
        End Function
        Private Sub SetPrimaryKey(ByVal entity As TEntity, ByVal key As TPrimaryKey) Implements IRepository(Of TEntity, TPrimaryKey).SetPrimaryKey
            SetPrimaryKeyCore(entity, key)
        End Sub
    End Class
End Namespace

File: CS/PersonalOrganizer/Common/DataModel/DesignTimeRepository.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using DevExpress.Mvvm;
using PersonalOrganizer.Common.Utils;

namespace PersonalOrganizer.Common.DataModel {
    public abstract class DesignTimeRepository<TEntity, TPrimaryKey> : DesignTimeReadOnlyRepository<TEntity>, IRepository<TEntity, TPrimaryKey>
        where TEntity : class {
        Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression;
        Func<TEntity, TPrimaryKey> getPrimaryKeyFunction;
        Action<TEntity, TPrimaryKey> setPrimaryKeyAction;

        public DesignTimeRepository(Expression<Func<TEntity, TPrimaryKey>> getPrimaryKeyExpression, Action<TEntity, TPrimaryKey> setPrimaryKeyAction = null) {
            this.getPrimaryKeyExpression = getPrimaryKeyExpression;
            this.setPrimaryKeyAction = setPrimaryKeyAction;
            if(this.setPrimaryKeyAction == null)
                this.setPrimaryKeyAction = getPrimaryKeyExpression.SetValueAction().Compile();
        }
        protected virtual TEntity CreateCore() {
            return DesignTimeHelper.CreateDesignTimeObject<TEntity>();
        }
        protected virtual TEntity FindCore(TPrimaryKey key) {
            throw new InvalidOperationException();
        }
        protected virtual void RemoveCore(TEntity entity) {
            throw new InvalidOperationException();
        }
        protected virtual TEntity ReloadCore(TEntity entity) {
            throw new InvalidOperationException();
        }
        protected virtual TPrimaryKey GetPrimaryKeyCore(TEntity entity) {
            if(getPrimaryKeyFunction == null)
                getPrimaryKeyFunction = getPrimaryKeyExpression.Compile();
            return getPrimaryKeyFunction(entity);
        }
        protected virtual void SetPrimaryKeyCore(TEntity entity, TPrimaryKey key) {
            setPrimaryKeyAction(entity, key);
        }
        #region IRepository
        TEntity IRepository<TEntity, TPrimaryKey>.Find(TPrimaryKey key) {
            return FindCore(key);
        }
        void IRepository<TEntity, TPrimaryKey>.Remove(TEntity entity) {
            RemoveCore(entity);
        }
        TEntity IRepository<TEntity, TPrimaryKey>.Create() {
            return CreateCore();
        }
        TEntity IRepository<TEntity, TPrimaryKey>.Reload(TEntity entity) {
            return ReloadCore(entity);
        }
        Expression<Func<TEntity, TPrimaryKey>> IRepository<TEntity, TPrimaryKey>.GetPrimaryKeyExpression {
            get { return getPrimaryKeyExpression; }
        }
        TPrimaryKey IRepository<TEntity, TPrimaryKey>.GetPrimaryKey(TEntity entity) {
            return GetPrimaryKeyCore(entity);
        }
        void IRepository<TEntity, TPrimaryKey>.SetPrimaryKey(TEntity entity, TPrimaryKey key) {
            SetPrimaryKeyCore(entity, key);
        }
        #endregion
    }
}